distributed/comm/mpi.py

```diff
@@ -0,0 +1,781 @@
+
+            local_rank = MPI.COMM_WORLD.Get_rank()
+
+            data = str(local_rank).encode()
+            writer.write(data)
+            await writer.drain()
+            logger.debug("listen(): wrote data")
+
+            assert reader.at_eof() == False 
+            data = await reader.read(4)
+            peer_rank = int(data.decode())
+            logger.debug("listen(): read data")
+
+            suggested_tag = tag_table.get(peer_rank)
+            #new_suggested_tag = suggested_tag + TAG_QUOTA_PER_CONNECTION
+            tag_table.update({peer_rank : (suggested_tag + TAG_QUOTA_PER_CONNECTION)})
+            logger.debug("%s: listen(): data=%s, rank=%s, peer_rank=%s, suggested_tag=%s", \
+                                                      socket.gethostname(), data, local_rank, peer_rank, suggested_tag)
+
+            mpi = MPI4Dask(
+                peer_rank,
+                suggested_tag,
+                local_addr = self.prefix + local_addr[0] + ":" + \
+                                                   str(local_addr[1]),
+                peer_addr = self.prefix + peer_addr[0] + ":" + \
+                                                   str(peer_addr[1]),
+                p_addr = peer_addr[0],
+                deserialize=self.deserialize,
+            )
+
+            mpi.allow_offload = self.allow_offload
+
+            try:
+                await self.on_connection(mpi)
+            except CommClosedError:
+                logger.debug("Connection closed before handshake completed")
+                return
+
+            if self.comm_handler:
+                logger.debug("%s: calling comm_handler(mpi)", socket.gethostname())
+                await self.comm_handler(mpi)
+
+        init_once()
+        logger.debug("asyncio.start_server()")
+        coro = asyncio.start_server(serve_forever, \
+                                    None, \
+                                    port=self._input_port, \
+                                    backlog=1024)
+        task = asyncio.create_task(coro)
+        self.mpi_server = await task
+        addr = self.mpi_server.sockets[0].getsockname()
+        logger.debug(f'Serving on {addr}')
+
+    def stop(self):
+        self.mpi_server = None
+
+    def get_host_port(self):
+        # TODO: TCP raises if this hasn't started yet.
+        return self.ip, self.port
+
+    @property
+    def listen_address(self):
+        return self.prefix + unparse_host_port(*self.get_host_port())
+
+    @property
+    def contact_address(self):
+        host, port = self.get_host_port()
+        host = ensure_concrete_host(host)  # TODO: ensure_concrete_host
+        return self.prefix + unparse_host_port(host, port)
+
+    @property
+    def bound_address(self):
+        # TODO: Does this become part of the base API? Kinda hazy, since
+        # we exclude in for inproc.
+        return self.get_host_port()
+
+class MPI4DaskBackend(Backend):
```

Comment:
Is it possible to create an abstract base class that handles MPI and TCP's needs? This should avoid the duplication here

Reply:
This duplicate comment is coming from UCX device. We can come to this after addressing other comments.